### PR TITLE
KAFKA-3812 State store locking is incorrect

### PR DESCRIPTION
**Pull request by dguy**
Move all state directory creation/locking/unlocking/cleaning to a single class. Don't release the channel until the lock is released. Refactor code to make use of new class

### CODE REVIEW

**Comment by dguy** [comment]
@guozhangwang @enothereska

**Comment by enothereska** [comment]
Ok, now we have separate locks per task ID. LGTM.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StateDirectory.java [hunk 1]
```diff
@@ -0,0 +1,186 @@
...
+                return pathname.isDirectory() && name.matches("\\d+_\\d+");
+            }
+        });
+    }
+
+    private FileChannel getFileChannel(final TaskId taskId, final Path lockPath) throws IOException {
+        final FileChannel channel = FileChannel.open(lockPath, StandardOpenOption.CREATE, StandardOpenOption.WRITE);
+        final FileChannel existing = channels.putIfAbsent(taskId, channel);
+        if (existing != null) {
+            channel.close();
```
Comment (ijuma) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/StateDirectory.java]: Why is it safe to close the channel here given the problem outlined in the JIRA? Wouldn't this `close` potentially release the lock held by `existing`? Or are the paths not the same? Relevant documentation text:

``` java
 * <p> On some systems, closing a channel releases all locks held by the Java
 * virtual machine on the underlying file regardless of whether the locks were
 * acquired via that channel or via another channel open on the same file.  It
 * is strongly recommended that, within a program, a unique channel be used to
 * acquire all locks on any given file.
```

Comment (dguy) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/StateDirectory.java]: Thanks @ijuma - i meant to change that. It doesn't need to be a `ConcurrentHashMap` either as there is a 1-to-1 mapping from `StreamThread` to `StateDirectory`. Updating now

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StateDirectory.java [hunk 2]
```diff
@@ -0,0 +1,184 @@
...
+        return stateDir.listFiles(new FileFilter() {
+            @Override
+            public boolean accept(final File pathname) {
+                final String name = pathname.getName();
+                return pathname.isDirectory() && name.matches("\\d+_\\d+");
+            }
+        });
+    }
+
+    private FileChannel getFileChannel(final TaskId taskId, final Path lockPath) throws IOException {
```
Comment (ijuma) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/StateDirectory.java]: Nitpick: I'd call this `getOrCreateFileChannel`.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StandbyTask.java [hunk 3]
```diff
@@ -38,15 +38,15 @@
 
     /**
      * Create {@link StandbyTask} with its assigned partitions
-     *
-     * @param id                    the ID of this task
+     *  @param id                    the ID of this task
```
Comment (guozhangwang) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/StandbyTask.java]: nit: one less space,

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StateDirectory.java [hunk 4]
```diff
@@ -0,0 +1,184 @@
...
+        }
+        stateDir = new File(baseDir, applicationId);
+        if (!stateDir.exists()) {
+            stateDir.mkdir();
+        }
+
+    }
+
+    /**
+     * Get ore create the directory for the {@link TaskId}
```
Comment (guozhangwang) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/StateDirectory.java]: nit: or

**Comment by guozhangwang** [comment]
One minor comment and otherwise LGTM.

Great cleaning up the unit tests as well!

**Comment by dguy** [comment]
Thanks @guozhangwang - comment fixed.

**Comment by guozhangwang** [comment]
Merged to trunk.

**Pull request closed without merging**